osquery/tables/events/darwin/yara.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ *  Copyright (c) 2015, Wesley Shields
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <map>
+#include <string>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+
+#include <yara.h>
+
+#include "osquery/events/darwin/fsevents.h"
+
+namespace osquery {
+namespace tables {
+
+/**
+ * @brief Track YARA matches to files.
+ */
+class YARAEventSubscriber : public EventSubscriber<FSEventsEventPublisher> {
+  DECLARE_SUBSCRIBER("yara");
+
+ public:
+  void init();
+
+ private:
+  // XXX: Is there a better way to say "I'm not ready to receive events"?
+  bool ready = false;
+  std::map<std::string, YR_RULES *> rules;
+
+  /**
+   * @brief This exports a single Callback for FSEventsEventPublisher events.
+   *
+   * @param ec The Callback type receives an EventContextRef substruct
+   * for the FSEventsEventPublisher declared in this EventSubscriber subclass.
+   *
+   * @return Status
+   */
+  Status Callback(const FSEventsEventContextRef& ec, const void* user_data);
+};
+
+/**
+ * @brief Each EventSubscriber must register itself so the init method is
+ * called.
+ *
+ * This registers YARAEventSubscriber into the osquery EventSubscriber
+ * pseudo-plugin registry.
+ */
+REGISTER(YARAEventSubscriber, "event_subscriber", "yara");
+
+/**
+ * The callback used when there are compilation problems in the rules.
+ */
+void YARACompilerCallback(int error_level,
+                          const char* file_name,
+                          int line_number,
+                          const char* message,
+                          void* user_data) {
+  if (error_level == YARA_ERROR_LEVEL_ERROR) {
+    VLOG(1) << file_name << "(" << line_number << "): error: " << message;
+  }
+  else {
+    VLOG(1) << file_name << "(" << line_number << "): warning: " << message;
+  }
+}
+
+void YARAEventSubscriber::init() {
+  YR_COMPILER *compiler = nullptr;
+  bool compiled;
+
+  int result = yr_initialize();
+  if (result != ERROR_SUCCESS) {
+    VLOG(1) << "Unable to initalize YARA.";
```

Comment:
want to emit the result error:
```cpp
VLOG(1) << "Unable to initialize YARA error: " << result;
```

I'd also suggest bumping that to `LOG(WARNING)`, then we can create a yara-wiki or GH issue to explain the log line.